src/mtype.c

```diff
@@ -3980,6 +3980,8 @@ void TypeSArray::resolve(Loc loc, Scope *sc, Expression **pe, Type **pt, Dsymbol
     }
     else
     {
+        if ((*pt)->ty != Terror)
+            next = *pt;     // prevent re-running semantic() on 'next'
```

Comment:
What I'm concerned about is much more meta. There doesn't seem to be a clear separation and encapsulation of processing in the compiler. It's become a mass of ad-hoc patches with no discernible organization. The symptom of this is the parade of regressions introduced by bug fixes. I'm afraid of it becoming completely intractable. We need to adopt a more principled approach. (The switch to ddmd will help a lot with this.)